[Feat] Apply polygon filter - Geo-Operations 2

macrigiuseppe opened the pull request:
Implements #390 

### Implementation details
#### Storing filters
I created a new property __layerId__ to store a reference to the mapped layer for the current polygon.
The polygon (feature) is stored in the filter __value__ property.
Feature filter is also deleted upon deletion of paired feature

#### Filtering
In order to filter out data correctly I integrated with [turf.js](https://turfjs.org/). In order to provide better flexibility and also be able to filter geojson elements I used [booleanWithin](https://turfjs.org/docs/#booleanWithin)
```javascript
import booleanWithin from '@turf/boolean-within';
```

#### Examples
![geo-ops-1](https://user-images.githubusercontent.com/2554552/62860351-3e09f380-bd00-11e9-98df-baa8a682d240.gif)

![geo-ops-2](https://user-images.githubusercontent.com/2554552/62860377-4e21d300-bd00-11e9-8b4f-c2e16e7b8254.gif)

<img width="1397" alt="Screen Shot 2019-12-10 at 4 25 38 PM" src="https://user-images.githubusercontent.com/2554552/70580774-d65a9c80-1b69-11ea-8c7d-9a7a6d634255.png">

Added Feature type to filter header

<img width="1792" alt="Screen Shot 2020-01-11 at 7 46 10 PM" src="https://user-images.githubusercontent.com/2554552/72213893-a0cc0a80-34ab-11ea-8eef-575e235113c1.png">

Enabled filter feature visibility toggle

![toggle-feature](https://user-images.githubusercontent.com/2554552/72213902-b2151700-34ab-11ea-9eb3-5f9fa4129570.gif)

Split Map
<img width="1363" alt="Screen Shot 2020-01-13 at 2 55 55 PM" src="https://user-images.githubusercontent.com/2554552/72299164-e3f5bd00-3614-11ea-9c61-d31ba8b0f98b.png">

![geo-filter-split-map](https://user-images.githubusercontent.com/2554552/72394279-7a47e280-36ea-11ea-9780-3b39b5e03200.gif)

All features
![apply-geo-filters](https://user-images.githubusercontent.com/2554552/72214136-2605ee00-34b1-11ea-9e59-0f5a370afe39.gif)

heshan0131 commented on src/utils/filter-utils.js:
```diff
@@ -99,6 +102,26 @@ export const TIME_ANIMATION_SPEED = [
+  return generateFilter({
+    dataId: layer.config.dataId,
+    layerId: layer.id,
+    fixedDomain: false,
```
I think Polygon filter should be `fixedDomain: true`, like time filter. It's weird when people select filter to a subset of point and its color changes.

heshan0131 commented on src/utils/filter-utils.js:
```diff
@@ -123,7 +146,10 @@ export function getDefaultFilter(dataId) {
+    interval: null,
+
+    // polygon:
+    layerId: null
```
I think it's time we create unique properties for each filter. instead of adding everything to default filter.  I imagine we can store not only layerId, but also layer.columns, and converted polygon geometry by turf in it.

heshan0131 commented on src/utils/filter-utils.js:
```diff
@@ -489,6 +513,20 @@ export function isInRange(val, domain) {
+ */
+export function isInPolygon(point, polygon) {
+  const convertedPoint = turfPoint(point);
+  const convertedPolygon = turfPolygon(polygon.geometry.coordinates);
```
Here you convert filter.value every time. should only need to do it once. you can store convertedPolygon in polygon filter?

heshan0131 replied:
wouldn't it be more performant to http://turfjs.org/docs/#booleanPointInPolygon?

macrigiuseppe replied:
> wouldn't it be more performant to http://turfjs.org/docs/#booleanPointInPolygon?

I am not sure booleanPointInPolygon is more performant. When we convert a point to a turf point, the underline representation of convertedPoint is a geojson object hence i am not sure about performance.
On the other hand booleanWithin is more flexible and we can pass any kind of shape (not just points) through geojson therefore we could use for different layers

macrigiuseppe commented:
@heshan0131 
New feature:
- Able to edit/move existing geo features:
![edit-feature](https://user-images.githubusercontent.com/2554552/63183223-cb2caf80-c054-11e9-83d2-745eb0c15bb0.gif)

Moving turfPolygon creation to filter creation code definitely improved performance

macrigiuseppe commented:
Filtering while modifying polygon
![on-the-fly-filter](https://user-images.githubusercontent.com/2554552/63278351-63ba6e00-c2a7-11e9-8d4c-fff72d20fca7.gif)

macrigiuseppe commented:
@heshan0131 Added new filter panel with data info:

<img width="1792" alt="Screen Shot 2019-08-29 at 4 48 45 PM" src="https://user-images.githubusercontent.com/2554552/63950847-08456880-ca7d-11e9-885c-c80fea525a98.png">

heshan0131 commented:
Created a polygon, go to select mode, select it, then click on rectangle icon, it tries to draw a rectangle but also filtering the data by the intersection

<img width="833" alt="Screen Shot 2019-09-04 at 4 34 01 PM" src="https://user-images.githubusercontent.com/3605556/64300504-dbd49500-cf31-11e9-9cc1-a5e9c44820f8.png">

heshan0131 commented:
When I delete a polygon by right click select 'delete' in the floating panel, The filter on the left side is still there. Once linked filter and the actual polygon should associate to each other, delete either will delete the other

<img width="1008" alt="Screen Shot 2019-09-04 at 4 52 46 PM" src="https://user-images.githubusercontent.com/3605556/64301589-88b01180-cf34-11e9-802f-db1a175d57e2.png">

heshan0131 commented on src/reducers/vis-state-updaters.js:
```diff
@@ -1328,3 +1344,167 @@ export function updateAllLayerDomainData(state, dataId, newFilter) {
+ * @param {Object} payload.layer
+ * @return {Object} nextState
+ */
+export function togglePolygonFilterUpdater(state, payload) {
```
do we need to store `feature` that is also a polygon filter in 2 places? It's making the logic really complicated. and you have to add this `filterData` everywhere in all feature update / delete calls.

I think as soon as someone toggle a polygon into a filter, you can move the `feature` into `polygonFilter.value`. So all the update / remove call on polygon filter can just be using `setFilter` `removeFilter`

Obviously, on the `Editor` side, you will have to assign different `onUpdate`, `onDeleteFeature` based on whether the `selectedFeatureId` is a filter or not. But I think this simplifies the logic in `visStateUpdater` so we don't add this many specific updaters just for polygon feature

macrigiuseppe replied:
in filter.value we store the turf.js representation of the feature and not original one so in case a user wants to apply the same feature as filter to multiple layers we still need the original one.
Another reason. If a user has a feature and a filter mapped against and decide to delete/toggle the filter, the feature should still be present on the map

heshan0131 replied:
Again on this point. 1. turf representation and the original feature is the same. 2. if a user deletes a polygon filter on map, the feature should not be present on the map. I don't think we should store them in 2 places. we would have to implement `getFilterIdxByFeatureId`, `currentFeature = editorFeatures.find(feat => feat.id === featureId);` just to get each one.

heshan0131 replied:
I tested out in [this commit](https://github.com/keplergl/kepler.gl/commit/7f927840ab976305eee3200d3aedb903ab60c0df#diff-3c42abf857bb61c5f6578c72771405d2R1658)

macrigiuseppe commented on src/reducers/vis-state-updaters.js:
```diff
@@ -270,6 +236,86 @@ function updateTextLabelPropAndValue(idx, prop, value, textLabel) {
+  }, []);
+}
+
+function removeFeatureFilters(state, filtersIdx) {
```
done

macrigiuseppe commented:
@heshan0131 Rebased with Master, adapted the logic to multiple dataset filters and added new tests.
I will work with the designer on the filter panel for geoperations

macrigiuseppe commented on test/fixtures/test-hex-id-data.js:
```diff
@@ -530,7 +530,13 @@ mergedH3Layer.config = {
   isVisible: true,
   highlightColor: [252, 242, 26, 255],
   isConfigActive: false,
-  colorField: fields[1],
```
because we are now loading layers before filters, fields are not updated with filterProps.

heshan0131 replied:
that's going to be a problem. we need field.mappedValue to calculate layer domain

macrigiuseppe commented on src/reducers/vis-state-updaters.js:
```diff
@@ -1007,8 +1042,8 @@ export const receiveMapConfigUpdater = (
   let mergedState = !keepExistingConfig ? resetMapConfigUpdater(state) : state;
-  mergedState = mergeFilters(mergedState, filters);
   mergedState = mergeLayers(mergedState, layers);
+  mergedState = mergeFilters(mergedState, filters);
```
in order for polygon filters to be loaded correctly we need load layers before filters

heshan0131 replied:
see above comment

heshan0131 commented on src/components/editor/index.js:
```diff
@@ -120,38 +131,59 @@ class Draw extends Component {
+        className={classnames('editor', className)}
+        style={style}
+      >
         <Editor
```
<img width="460" alt="Screen Shot 2019-12-24 at 5 46 19 AM" src="https://user-images.githubusercontent.com/3605556/71381966-dac69100-2610-11ea-9ac6-e03b9a26b3fa.png">

Can you create a task in upgrade react-map-gl-drawer so this doesn't showup?

macrigiuseppe replied:
Done: https://github.com/keplergl/kepler.gl/issues/877

heshan0131 commented on src/components/editor/index.js:
```diff
@@ -120,38 +131,59 @@ class Draw extends Component {
+
     const {selectedFeature = {}} = editor;
+    const {lastPosition, showActions} = this.state;
+    const selectedFeatureId = (selectedFeature || {}).id;
```
`{selectedFeature = {}} = editor` is doing the same thing. No need to do it again here.

macrigiuseppe replied:
that's true

heshan0131 commented on src/components/editor/index.js:
```diff
@@ -120,38 +131,59 @@ class Draw extends Component {
     const {selectedFeature = {}} = editor;
+    const {lastPosition, showActions} = this.state;
+    const selectedFeatureId = (selectedFeature || {}).id;
+    const currentFilter = filters.find(f => f.value.id === selectedFeatureId);
```
is it possible where `f.value` is not defined?

heshan0131 commented:
nitpick
<img width="1039" alt="Screen Shot 2019-12-24 at 6 12 59 AM" src="https://user-images.githubusercontent.com/3605556/71383230-49f2b400-2616-11ea-804f-5df44b90b0f6.png">
when toolbar is horizontal, the border-right is 1. when it is vertical, the border-right should be 0

heshan0131 commented on src/components/common/action-panel.js:
```diff
@@ -102,35 +103,49 @@ export const ActionPanelItem = React.memo(({
+      ) : (
+        <span className="label">{label}</span>
+      )}
+      {children && children.length ? ( <div className="label-icon">
```
why not wrap them in 1 div
```js
     {children && children.length ?
      <div>
        <div className="label-icon">
          <ArrowRight height="16px" />
        </div>
        <div className="nested-group">
        {React.Children.map(children, renderChildren)}
        </div>
        </div>
      : null}
```

heshan0131 commented on src/components/common/action-panel.js:
```diff
@@ -102,35 +103,49 @@ export const ActionPanelItem = React.memo(({
+  }, [onClick]);
+
+  return (
+    <StyledItem
```
Also nitpick:
the `:last` item should have `border-bottom: 0`

heshan0131 commented on src/components/common/field-selector.js:
```diff
@@ -19,26 +19,40 @@
       ) : null}
-      <span className={classList.listItemAnchor}>{displayOption(value)}</span>
-    </div>
+      <span className={classnames(classList.listItemAnchor, value.type, displayOption(value))}>
```
is it neccessary to add `displayOption(value)` to `className`?  what if it is a very long string

macrigiuseppe replied:
I can actually remove those, they may have been leftovers from previous versions

heshan0131 commented on src/components/common/item-selector/item-selector.js:
```diff
@@ -31,25 +31,7 @@ import Typeahead from './typeahead';
-
-  return [item];
-}
+import {toArray} from 'utils/utils';
```
nice

heshan0131 commented on src/components/common/styled-components.js:
```diff
@@ -81,6 +83,17 @@ export const PanelLabel = styled.label.attrs({
   text-transform: capitalize;
 `;
 
+export const PanelValue = styled.label.attrs({
```
is this used anywhere?

macrigiuseppe replied:
leftover and removed

heshan0131 commented on src/components/common/styled-components.js:
```diff
@@ -540,3 +553,33 @@ export const MapControlButton = styled.div.attrs({
   }
 `;
+
+export const StyledFilterPanel = styled.div`
```
`StyledFilterPanel` is also defined in `filters/components` and `side-panel/filter-panel/filter-panel`.  Can you delete duplicates and keep them all in one place? I don't think this should be a generic `styled-components` I would keep it under `side-panel`

macrigiuseppe replied:
i removed the code from here. I think i have started creating some generic components and stopped at some point

heshan0131 commented on src/components/common/styled-components.js:
```diff
@@ -540,3 +553,33 @@ export const MapControlButton = styled.div.attrs({
+  }
+`;
+
+export const StyledFilterHeader = styled(StyledPanelHeader)`
```
this is only used in `filter-panel-header`, if it's not used in other places, I would keep it there

heshan0131 commented on src/components/editor/feature-action-panel.js:
```diff
@@ -58,9 +60,13 @@ export class FeatureActionPanel extends PureComponent {
+    const {layerId = []} = (currentFilter || {});
+
     return (
       <StyledActionsLayer className={className} position={position}>
```
instead of passing this to styled-component. this should just be in component `style={{top: x, left: y}}` handled by React. It is not recommended to pass props that constantly change to styled-component, because it needs to inject to css whenever prop changes.

heshan0131 commented on src/components/editor/index.js:
```diff
@@ -44,17 +44,28 @@ const StyledWrapper = styled.div`
   cursor: ${props => props.editor.mode === EDITOR_MODES.EDIT ? 'pointer' : 'crosshair'};
 `;
 
+const AVAILABLE_LAYERS = [
```
Shouldn't this be in `default-settings`?

macrigiuseppe replied:
I will move it to default-settings but i will rename it into
```
EDITOR_AVAILABLE_LAYERS
```

heshan0131 commented on src/components/filters/filter-panels/range-filter-panel.js:
```diff
@@ -0,0 +1,112 @@
+      [idx, setFilter]);
+
+    return (
+      <>
```
should this be a <div>?

heshan0131 commented on src/utils/filter-utils.js:
```diff
@@ -369,6 +435,74 @@ export function filterData(dataset, filters) {
+  },
+  // layers contain only layers for the current dataset
+  [FILTER_TYPES.polygon]: (data, filter, index, field, layers) => {
+    if (!(layers || layers.length === 0)) {
```
This function is called on every data point, `currentLayers` should be passed into this function. Instead of creating it on every pass.

heshan0131 commented on src/utils/filter-utils.js:
```diff
@@ -369,6 +435,74 @@ export function filterData(dataset, filters) {
+    ).filter(l => Boolean(l));
+
+    return currentLayers.every(layer => {
+      const points = getLayerPointsToFilter(layer, data);
```
Each layer has a method `getPositionAccessor` which will return a position accessor, you can get point position. no need for the `getLayerPointsToFilter` function
```
const getPosition = layer.getPositionAccessor();
const points = getPosition(data);
```

heshan0131 commented on src/utils/filter-utils.js:
```diff
@@ -309,7 +375,7 @@ export function getFieldDomain(data, field) {
  * @param {Object[]} filters list of filters to use against dataset
  */
-export function filterData(dataset, filters) {
+export function filterData(dataset, filters, layers) {
```
in this function, we can create a `filterFunctors(field, dataset.id, filter, layers)` instead of a `filtersToFields` with all the pre compyting before passing data point to it.  filterFunction is a function that takes `(data, index) => true || false`, it shouldn't does any computation rather than get the value from data and compare it to filter.value. which means by passing `field, dataset.id, filter, layers` to `getFilterFunction` we on longer need to pass it to `isDataMatchFilter`

implementation as below:

```js
  const appliedFilters = filters.filter(d => shouldApplyFilter(d, dataset.id));

  // Map filter against current dataset field
  const filterFunctors = appliedFilters.reduce((acc, filter) => {
    const fieldIndex = getDatasetFieldIndexForFilter(dataset, filter);
    const field = fieldIndex !== -1 ? fields[fieldIndex] : null;

    return {
      ...acc,
      [filter.id]: getFilterFunction(field, dataset.id, filter, layers)
    }
  }, {});


function getFilterFunction(field, dataId, filter, layers) {
  // field could be null
  let valueAccessor = data => field ? data[field.tableFieldIndex - 1] : null;

  switch (filter.type) {
    case FILTER_TYPES.range:
      return data => {
        return isInRange(valueAccessor(data), filter.value)
      };
    case FILTER_TYPES.multiSelect:
      return data => filter.value.includes(valueAccessor(data));
    case FILTER_TYPES.select:
      return data => valueAccessor(data) === filter.value;
    case FILTER_TYPES.timeRange:
      if (field && field.filterProp && Array.isArray(field.filterProp.mappedValue)) {
        valueAccessor = (data, index) => field.filterProp.mappedValue[index];
      } else {
        valueAccessor = data => timeToUnixMilli(valueAccessor(data), field.format);
      }
      return (data, index) => isInRange(valueAccessor(data, index), filter.value);
    case FILTER_TYPES.polygon:
      if (!(layers || layers.length === 0)) {
        return data => true;
      }

      const layerFilterFunctions = filter.layerId
        .map(id => layers.find(l => l.id === id))
        .filter(l => l && l.config.dataId === dataId)
        .map(layer => getPolygonFilterFunctor(layer, filter));

      return data => layerFilterFunctions.every(filterFunc => filterFunc(data));
    default:
      return data => true;
  }
};
```

macrigiuseppe replied:
this looks way better, thank you

macrigiuseppe replied:
had to change few things in the above code because it was causing the following error:
```
not ok 1 RangeError: Maximum call stack size exceeded
  ---
    operator: error
    expected: |-
      undefined
    actual: |-
      [RangeError: Maximum call stack size exceeded]
    at: Test.die (/Users/gmacri/Documents/projects/kepler.gl/node_modules/tape-catch/index.js:56:10)
    stack: |-
      RangeError: Maximum call stack size exceeded
          at valueAccessor (/Users/gmacri/Documents/projects/kepler.gl/src/utils/filter-utils.js:418:25)
          at valueAccessor (/Users/gmacri/Documents/projects/kepler.gl/src/utils/filter-utils.js:418:49)
```
@ line: ```valueAccessor = data => timeToUnixMilli(valueAccessor(data), field.format);```

The overall result is way better

heshan0131 commented on src/utils/filter-utils.js:
```diff
@@ -344,15 +410,15 @@ export function filterData(dataset, filters) {
       // filter data used to calculate layer Domain
       const matchForDomain = dynamicDomainFilters.every(filter => {
-        return isDataMatchFilter(d, filter, i, filtersToFields[filter.id]);
+        return isDataMatchFilter(d, filter, i, filtersToFields[filter.id], layers);
```
no longer needed, simply call 
```js
 const matchForDomain = dynamicDomainFilters
        .every(filter => filterFunctors[filter.id](d, i));
```

heshan0131 commented on src/utils/filter-utils.js:
```diff
@@ -344,15 +410,15 @@ export function filterData(dataset, filters) {
         // filter data for render
         const matchForRender = fixedDomainFilters.every(filter =>
-          isDataMatchFilter(d, filter, i, filtersToFields[filter.id])
+          isDataMatchFilter(d, filter, i, filtersToFields[filter.id], layers)
```
no longer needed, simply call 
```js
 const matchForDomain = dynamicDomainFilters
        .every(filter => filterFunctors[filter.id](d, i));
```

macrigiuseppe replied:
```
const matchForRender = fixedDomainFilters
        .every(filter => filterFunctors[filter.id](d, i));
```

heshan0131 commented on src/utils/filter-utils.js:
```diff
@@ -369,6 +435,74 @@ export function filterData(dataset, filters) {
+ * @param data
+ * @return {*[][]}
+ */
+export const getLayerPointsToFilter = (layer, data) => {
```
each layer has a `.getPositionAccessor` method, here you can precalculate `getPosition` and pass data to it

```js
export const getPolygonFilterFunctor = (layer, filter) => {
  const getPosition = layer.getPositionAccessor();

  switch (layer.type) {
    case LAYER_TYPES.arc:
    case LAYER_TYPES.line:
      return data => {
        const pos = getPosition({data});
        return pos.every(Number.isFinite) && [
          [pos[0], pos[1]],
          [pos[3], pos[4]]
        ].every(point => isInPolygon(point, filter.value));
      }
    default:
      return data => {
        const pos = getPosition({data});
        return pos.every(Number.isFinite) && isInPolygon(pos, filter.value);
      }
  }
};
```

heshan0131 commented on src/utils/filter-utils.js:
```diff
@@ -379,32 +513,9 @@ export function filterData(dataset, filters) {
-    default:
-      return true;
-  }
+export function isDataMatchFilter(data, filter, i, field, layers = null) {
```
after mapping filter to `filterFunction`, we dont' need this function anumore

macrigiuseppe commented:
> nitpick
> <img alt="Screen Shot 2019-12-24 at 6 12 59 AM" width="1039" src="https://user-images.githubusercontent.com/3605556/71383230-49f2b400-2616-11ea-804f-5df44b90b0f6.png">
> when toolbar is horizontal, the border-right is 1. when it is vertical, the border-right should be 0

<img width="926" alt="Screen Shot 2019-12-24 at 5 47 09 PM" src="https://user-images.githubusercontent.com/2554552/71420411-6e31ad80-2675-11ea-8105-f6130bda9376.png">

heshan0131 commented on src/components/map-container.js:
```diff
@@ -468,22 +479,20 @@ export default function MapContainerFactory(MapPopover, MapControl) {
+              editor={editor}
+              filters={this.polygonFilters(this.props)}
               isEnabled={isEdit}
               layers={layers}
```
Should we show layer that is currently hidden? It's probably better to only show visible layer. You can achieve that by passing `layersToRender`, it is a `{[layer.id]: layer}` Object

macrigiuseppe replied:
layersToRender container {[layer.id]: true|false}.
It looks like the current pattern is to pass both __layers__ and __layersToRender__ so I need to pass both layers and layers to render

heshan0131 commented on src/processors/data-processor.js:
```diff
@@ -203,6 +203,7 @@ function cleanUpFalsyCsvValue(rows) {
  * Process uploaded csv file to parse value by field type
  *
  * @param {Array<Array>} rows
+ * @param geoFieldIdx geo field index
```
it should be {Number}

heshan0131 commented on src/reducers/vis-state-updaters.js:
```diff
@@ -111,12 +120,19 @@ disableStackCapturing();
 /* eslint-enable no-unused-vars */
 
-export const defaultAnimationConfig = {
+export const defaultAnimationConfig ={
```
missing space

heshan0131 commented on src/reducers/vis-state-updaters.js:
```diff
@@ -183,10 +199,8 @@ export const INITIAL_VIS_STATE = {
-    features: []
-  }
+  editor: DEFAULT_EDITOR
+
```
empty line

heshan0131 commented on src/reducers/vis-state-updaters.js:
```diff
@@ -535,8 +553,8 @@ export function setFilterUpdater(state, action) {
-  // ENsuring backward compatibility
-  const datasetIds = Array.isArray(dataId) ? dataId : [dataId];
+  // Ensuring backward compatibility
+  let datasetIds = Array.isArray(dataId) ? dataId : [dataId];
```
use `toArray`?

heshan0131 commented on src/reducers/vis-state-updaters.js:
```diff
@@ -799,10 +830,18 @@ export const removeFilterUpdater = (state, action) => {
+      ...state.editor,
+      features: filter.type !== FILTER_TYPES.polygon ?
+        state.editor.features : state.editor.features.filter(f => f.id !== filter.value.id),
+      selectedFeature: null
```
if `filter.type !== FILTER_TYPES.polygon ` should `selectedFeature` be set to null?
Looks like you only need to update editor if ``filter.type === FILTER_TYPES.polygon` so why not
```js

  const newState = {
    ...state,
    datasets: applyFiltersToDatasets(dataId, state.datasets, newFilters, state.layers),
    filters: newFilters,
    editor:  filter.type === FILTER_TYPES.polygon ? {
      ...state.editor,
      features: state.editor.features.filter(f => f.id !== filter.value.id),
      selectedFeature: null
    } : state.editor
  };
```

heshan0131 commented on src/reducers/vis-state-updaters.js:
```diff
@@ -1524,3 +1530,215 @@ export function updateAnimationDomain(state) {
+  const newFilters = Object.assign([...state.filters], {[filtersIdx]: newFilter});
+
+  // no filters found
+  if (filtersIdx === -1) {
```
shouldn't this be moved to right after `const filtersIdx = getFilterIdxByFeatureId(newState, selectedFeature.id);`

heshan0131 commented on src/reducers/vis-state-updaters.js:
```diff
@@ -1524,3 +1530,215 @@ export function updateAnimationDomain(state) {
+    editor: {
+      ...state.editor,
+      features,
+      mode: lastFeature && lastFeature.properties.isClosed ?
```
what is `isClosed`? and why when `isClosed` is true editor mode should be EDIT?

macrigiuseppe replied:
isClosed is a property that comes from react-map-gl-draw. It means the feature we are currently drawing is completed, e.g. while drawing a polygon we finally draw all lines.

The abose logic allows the app to detect when the current feature is being completed and switch from POLYGON/RECTANGLE mode to EDIT mode so users can immediately select the latest created feature.

![select-feature](https://user-images.githubusercontent.com/2554552/71479625-731d6b00-27f5-11ea-9fff-bde963bcdf22.gif)

heshan0131 commented on src/reducers/vis-state-updaters.js:
```diff
@@ -1524,3 +1530,215 @@ export function updateAnimationDomain(state) {
+    return newState;
+  }
+
+  const filtersIdx = getFilterIdxByFeatureId(newState, selectedFeature.id);
```
shouldn't this be `filterIdx`?

heshan0131 commented on src/utils/filter-utils.js:
```diff
@@ -840,3 +969,51 @@ export function getDatasetFieldIndexForFilter(dataset, filter) {
+ * @return {object} Filter
+ */
+export function updatePolygonFilter(filter, feature) {
+  const polygon = turfPolygon(feature.geometry.coordinates);
```
a turfPolygon is a geojson Polygon feature, same as the `feature` object react-map-gl-draw returns. Here you dn't need to call  `turfPolygon` and merge them. this works:
```js
export function updatePolygonFilter(filter, feature) {
  // const polygon = turfPolygon(feature.geometry.coordinates);
  return {
    ...filter,
    // we merge both turf and feature properties into one
    value: {
      id: feature.id,
      ...feature
    }
  }
}
```

heshan0131 commented on src/reducers/vis-state-merger.js:
```diff
@@ -111,13 +110,24 @@ export function mergeFilters(state, filtersToMerge) {
   const datasetsToFilter = uniq(flattenDeep(merged.map(f => f.dataId)));
 
-  const filtered = applyFiltersToDatasets(datasetsToFilter, updatedDatasets, updatedFilters);
+  const filtered = applyFiltersToDatasets(datasetsToFilter, updatedDatasets, updatedFilters, state.layers);
```
Can you add a test case to `vis-state-merger-test.js` with a map saved with a polygon filter, some other filter and a layer with `colorField`, then loaded it back to test this?

heshan0131 commented on src/actions/vis-state-actions.js:
```diff
@@ -508,6 +508,76 @@ export function setFeatures(features) {
+ * @param {string} feature.selectedFeatureId
+ * @return {{selectedFeatureId: *, type: ActionTypes.SET_SELECTED_FEATURE}}
+ */
+export function setSelectedFeature({selectedFeatureId}) {
```
Seems like all the places where `onSelect` are called in `editor.js`, you are only passing `{selectedFeatureId: null | id}`,  why not unwrap the argument to be `selectedFeatureId`?

macrigiuseppe replied:
Initially I thought I was going to propagate the full feature

heshan0131 commented on src/components/common/vertical-toolbar.js:
```diff
@@ -0,0 +1,21 @@
+  padding: 0 24px;
+  
+  .toolbar-item {
+    padding: 16px;
```
vertical toolbar `toolbar-item` padding should be `16px 0`. Because toolbar already has side padding

heshan0131 commented on src/components/map/map-control.js:
```diff
@@ -240,16 +242,22 @@ const Toggle3dButton = React.memo(({dragRotate, onTogglePerspective}) => (
 `;
 
-const MapDrawPanel = React.memo(({editor, isActive, onToggleMenuPanel, onSetEditorMode}) => {
+const MapDrawPanel = React.memo(({
```
In split map, is it possible to only show `MapDrawPanel` in the split map clicked?  Currently, it is shown in both, which is a little weird.
e,g, in `mapControls.mapDraw` we can save the `splitMap` index [true, false], instead of `true | false`. However, this means in `ui-state` reducer has to listen to `TOGGLE_SPLIT_MAP` and update mapDraw

<img width="966" alt="Screen Shot 2020-01-07 at 9 18 34 PM" src="https://user-images.githubusercontent.com/3605556/71953746-fc3e8580-3197-11ea-95ad-72cc62849d15.png">

macrigiuseppe replied:
We may have a similar problem with the legend being displayed in both maps at the sam time
hmm do you think a lot of people will be using the editor mode in split map view?
We could disable the editor when in split mode

heshan0131 replied:
displaying legend and map control panel in 2 places is fine actually. It allows people to compare whats different in 2 split map. But allow drawying in split maps is useful. you can imagine people want to draw in one map, and see how the other map is filtered. 

For example. showing begin trip in one map, and dropoff in another. Drawing in begin trip to see how dropoff map changes...

macrigiuseppe replied:
ok, that makes sense

heshan0131 commented on src/reducers/vis-state-updaters.js:
```diff
@@ -311,6 +325,10 @@ function updateTextLabelPropAndValue(idx, prop, value, textLabel) {
 }
 
+function getFilterIdxByFeatureId(state, featureId) {
+  return state.filters.findIndex(f => f.type === FILTER_TYPES.polygon && f.value.id === featureId);
```
what if `f.value` is `null`?

heshan0131 commented on src/reducers/vis-state-merger.js:
```diff
@@ -111,13 +110,24 @@ export function mergeFilters(state, filtersToMerge) {
     filters: updatedFilters,
     datasets: filtered,
-    filterToBeMerged: unmerged
+    filterToBeMerged: unmerged,
```
now that we merge filter twice, `filterToBeMerged` should be `[...state.filterToBeMerged, ...unmerged]`

macrigiuseppe replied:
this change is breaking a lot of tests. need to dig a little bit

macrigiuseppe replied:
this may not work because if we consider the following scenario:
- a filter is added to filterToBeMerged list
- a new dataset is added that matches such filter, therefore, we try to merge filters again
- when we merge filter, ```state.filterToBeMerged``` has one element and unmerged has 0, therefore, the new state ```filterToBeMerged``` property will have again one element because of ```[...state.filterToBeMerged, ...unmerged]```

macrigiuseppe replied:
@heshan0131 ^

heshan0131 replied:
in cases where filterA is regular filter, filterB is layer based filter.  both are waiting for the dataset to be loaded. when we run mergeFilter once, we save filterA to unmerged. and when we run it again, we clear out unmerged, and save filterB to unmerged. It's wrong, because we are losing filterA

macrigiuseppe replied:
got it. Need to review the logic and understand why a lot of tests are failing

heshan0131 commented on src/reducers/vis-state-updaters.js:
```diff
@@ -1514,3 +1526,215 @@ export function updateAnimationDomain(state) {
+    filters: newFilters
+  };
+
+  return updateAllLayerDomainData(newState, Object.keys(newState.datasets));
```
dataId here should be `newFilter.dataId`, and you should pass `newFilter` to this function as well.
Because we only need to update the layer domains / layerdata associated with newly filtered dataset. and skipped the layers haven't been affected by it's parent datasets.

Also if the filter has fixedDomain, we don't need to update layer.domain, only need to update layerData

```
return updateAllLayerDomainData(newState, newFilter.dataId, newFilter);
```

macrigiuseppe replied:
no longer needed because we are now using setFilterUpdater

heshan0131 commented on src/reducers/vis-state-updaters.js:
```diff
@@ -1514,3 +1526,215 @@ export function updateAnimationDomain(state) {
+
+  const filter = state.filters[filterIdx];
+
+  const newFilter = updatePolygonFilter(filter, feature);
```
updatePolygonFilter in a nutshell updates filter.value, and all the code following it is to apply the filter.
You should be able to call  `setFilterUpdater` here with updated value, instead of all the lines below.
```
const featureInValue = {...feature, id: feature.id};
return setFilterUpdater(newState, {idx: filterIdx, prop: 'value', value: featureInValue})
```

macrigiuseppe replied:
actually it turns out if we call setFilterUpdater we don't need to use updatePolygonFilter and we can simply do the following:
```
return setFilterUpdater(newState, {idx: filterIdx, prop: 'value', value: feature});
```

macrigiuseppe replied:
no longer valid

heshan0131 commented on src/reducers/vis-state-updaters.js:
```diff
@@ -1514,3 +1526,215 @@ export function updateAnimationDomain(state) {
+  const {layer, featureId} = payload;
+  const {dataId} = layer.config;
+  const filterIdx = getFilterIdxByFeatureId(state, featureId);
+  let datasetIds = toArray(dataId);
```
you don't need `toArray` here because layer.config.dataId is always a single id, layer is only associated with 1 dataset.

heshan0131 commented on src/utils/filter-utils.js:
```diff
@@ -18,13 +18,18 @@
+import booleanWithin from '@turf/boolean-within';
+import {point as turfPoint} from '@turf/helpers';
 import {ALL_FIELD_TYPES} from 'constants/default-settings';
 import {maybeToDate, notNullorUndefined, unique} from './data-utils';
```
add `timeToUnixMilli` here

heshan0131 commented on src/utils/filter-utils.js:
```diff
@@ -304,12 +368,74 @@ export function getFieldDomain(data, field) {
+        ].every(point => isInPolygon(point, filter.value));
+      };
+    default:
+      return data => {
```
not all getPosition from layer returns point, i think for other layer type we should simply return `data => true`

heshan0131 commented on src/utils/filter-utils.js:
```diff
@@ -304,12 +368,74 @@ export function getFieldDomain(data, field) {
+    case FILTER_TYPES.timeRange:
+      const mapppedValue = get(field, ['filterProp', 'mappedValue']);
+      const accessor = Array.isArray(mapppedValue) ?
+        (data, index) => field.filterProp.mappedValue[index]
```
You already have `mapppedValue` defined
`(data, index) => mappedValue[index] : data => timeToUnixMilli(valueAccessor(data), field.format);`

heshan0131 commented on src/actions/vis-state-actions.js:
```diff
@@ -508,6 +508,76 @@ export function setFeatures(features) {
+ * @param {Object} layer
+ * @return {{feature: *, type: ActionTypes.TOGGLE_POLYGON_FILTER, layer: *}}
+ */
+export function togglePolygonFilter(layer, featureId) {
```
the naming of this action is a bit blurry. it sounds like this is to hide and show polygon filter. If layer is passed along can it be `setPolygonFilterLayer` ?

heshan0131 commented on src/reducers/vis-state-updaters.js:
```diff
@@ -1514,3 +1526,215 @@ export function updateAnimationDomain(state) {
+ * @param {[Object]} features to store
+ * @return {Object} nextState
+ */
+export function setFeaturesUpdater(state, {features = []}) {
```
regarding to the comments in this function
see [this commit](https://github.com/keplergl/kepler.gl/commit/7f927840ab976305eee3200d3aedb903ab60c0df#diff-3c42abf857bb61c5f6578c72771405d2R1556)

heshan0131 commented on src/schemas/vis-state-schema.js:
```diff
@@ -648,7 +648,11 @@ export const filterPropsV1 = {
     }
-  })
+  }),
+  fixedDomain: null,
```
fixedDomain is a property generated from `getDefaultFilter` we don't need to save it to the config

heshan0131 replied:
ping

heshan0131 commented on src/reducers/vis-state-updaters.js:
```diff
@@ -1514,3 +1526,215 @@ export function updateAnimationDomain(state) {
+ * @return {Object} nextState
+ */
+export function deleteFeatureUpdater(state, {featureId}) {
+  if (!featureId) {
```
pass feature to be deleted instead of `featureId`,
https://github.com/keplergl/kepler.gl/commit/7f927840ab976305eee3200d3aedb903ab60c0df#diff-3c42abf857bb61c5f6578c72771405d2R1615

macrigiuseppe replied:
the main reason is because react-map-gl-draw provide a feature.id when a selection is performed

heshan0131 commented on src/reducers/vis-state-updaters.js:
```diff
@@ -1514,3 +1526,215 @@ export function updateAnimationDomain(state) {
+ * @param {[Object]} features to store
+ * @return {Object} nextState
+ */
+export const setSelectedFeatureUpdater = (state, {selectedFeatureId}) => ({
```
instead of save the id in reducer, why not save the entire feature, that way it easier to literate over its property, also nicer to pass the entire feature around when deleting it
https://github.com/keplergl/kepler.gl/commit/7f927840ab976305eee3200d3aedb903ab60c0df#diff-3c42abf857bb61c5f6578c72771405d2R1600

heshan0131 commented on src/components/editor/index.js:
```diff
@@ -24,7 +24,8 @@ import styled from 'styled-components';
 import {Editor} from 'react-map-gl-draw';
```
also made some changes to the Event key
https://github.com/keplergl/kepler.gl/commit/7f927840ab976305eee3200d3aedb903ab60c0df#diff-37098adfde85b215e32b27850c960d0fR43

macrigiuseppe commented:
@heshan0131 
I added polygon type in filter panel header. I added the prefix __Geo__ because the previous one (Polygon) would have created a non optimal user experience since for non-rectangle features we would have displayed ```Polygon - Polygon```. See below

<img width="1792" alt="Screen Shot 2020-01-11 at 7 46 10 PM" src="https://user-images.githubusercontent.com/2554552/72220926-b7557e80-350a-11ea-899f-83e38d203089.png">

Added toggle visibility for filter feature. Initially, I was going to rename onToggleAnimation to something more generic to be reused in different filters but I think it's better to have different actions in case we want to extend filter panel in the future so I decided to create a new __toggleFilterFeature__ action. See below

![apply-geo-filters](https://user-images.githubusercontent.com/2554552/72220956-04395500-350b-11ea-84c2-8290717efa0a.gif)

heshan0131 commented on test/node/schemas/vis-state-schema-test.js:
```diff
@@ -102,7 +102,8 @@ test('#visStateSchema -> v1 -> save load filters', t => {
     plotType: 'histogram',
-    yAxis: null
+    yAxis: null,
+    fixedDomain: true
```
don't need to save `fixedDomain` prop

heshan0131 commented on src/utils/filter-utils.js:
```diff
@@ -134,33 +143,66 @@ export function getDefaultFilter(dataId) {
+    filter: {
+      ...filter,
+      freeze: true,
+      fieldIdx: [0]
```
why `fieldIdx` needs to be [0]? if there is no field selected, can't it be `[]`?

heshan0131 commented on src/components/filters/filter-panels/multi-select-filter-panel.js:
```diff
@@ -0,0 +1,112 @@
+      [idx, setFilter]);
+
+    return (
+      <>
```
<div>?

heshan0131 commented on src/components/filters/filter-panels/new-filter-panel.js:
```diff
@@ -0,0 +1,94 @@
+    );
+
+    return (
+      <>
```
<div className="new-filter-panel">

heshan0131 replied:
`<div className="new-filter-panel">`

heshan0131 commented on src/components/filters/filter-panels/polygon-filter-panel.js:
```diff
@@ -0,0 +1,102 @@
+    const featureType = get(filter, ['value', 'properties', 'renderType'], true);
+
+    return (
+      <>
```
`<div className="polygon-filter-panel">

macrigiuseppe replied:
i assigned ```<div className="polygon-filter-panel">```

heshan0131 commented on src/components/filters/filter-panels/range-filter-panel.js:
```diff
@@ -0,0 +1,112 @@
+      [idx, setFilter]);
+
+    return (
+      <div>
```
`<div className="range-filter-panel">`

heshan0131 commented on src/components/filters/filter-panels/single-select-filter-panel.js:
```diff
@@ -0,0 +1,113 @@
+      [idx, setFilter]);
+
+    return (
+      <>
```
same here

heshan0131 commented on src/components/filters/filter-panels/range-filter-panel.js:
```diff
@@ -0,0 +1,112 @@
+
+    return (
+      <div>
+        <FilterPanelHeader
```
```js
const onFieldSelector = useCallback(field =>
        setFilter(idx, 'name', field.name),
      [idx, setFilter]
    );

    const onSourceDataSelector = useCallback(value =>
        setFilter(idx, 'dataId', value),
      [idx, setFilter]
    );

    const onSetFilter = useCallback(value =>
        setFilter(idx, 'value', value),
      [idx, setFilter]);

<PanelHeader>
   <FieldSelector/>
</PanelHeader>
<StyledFilterContent>
    <SourceDataSelecto/>
</StyledFilterContent>
```

Has been repeated in every `xx-filter-panel` component, except polygon filter. Repeating them 5 times, doesn't make sense. Why not make a component `FilterPanelWithFieldSelect` and use it to wrap other filters?

macrigiuseppe replied:
I think I decided to go down this way because both polygon and time range filters have panel actions and by simply using children couldn't be recreated but I think we can create the common component and
- panel actions: will be passed as an array of object: `
``{onClick: enlargeFilter, tooltip:: 'Time Playback, iconComponent: Clock, onSelect: 
onFieldSelector}```
- the actual field component, e.g.TimeRangeFilter, will be passed as one of the children

macrigiuseppe commented:
Fixed map control padding:
<img width="1392" alt="Screen Shot 2020-01-17 at 2 13 11 PM" src="https://user-images.githubusercontent.com/2554552/72649827-87a3dd80-3933-11ea-84e7-cb3dc5b5533d.png">

Fixed map export
<img width="1197" alt="Screen Shot 2020-01-17 at 2 16 04 PM" src="https://user-images.githubusercontent.com/2554552/72649984-ee28fb80-3933-11ea-9dcc-1b558eb6785a.png">

heshan0131 commented on src/actions/vis-state-actions.js:
```diff
@@ -378,6 +378,18 @@ export function enlargeFilter(idx) {
+/**
+ * SHow/hide filter feature on map
+ * @param {Number} idx - index of filter feature to show/hide
+ * @return {{type: ActionTypes.TOGGLE_FEATURE_LAYER, idx: idx}}
```
return `TOGGLE_FILTER_FEATURE`

heshan0131 commented on src/actions/vis-state-actions.js:
```diff
@@ -508,6 +520,76 @@ export function setFeatures(features) {
+ * Set the current feature to be edited/deleted
+ * @memberof visStateActions
+ * @param {Object} feature
+ * @param {string} feature.selectedFeatureId
```
remove `@param {string} feature.selectedFeatureId`

heshan0131 commented on src/actions/vis-state-actions.js:
```diff
@@ -508,6 +520,76 @@ export function setFeatures(features) {
+ * It will apply the provide feature as filter to the given layer.
+ * If the given feature is already applied as filter to the layer it  will remove it from  the filter list
+ * @memberof visStateActions
+ * @param {Object} featureId feature to use as filter base for layer
```
JSDoc param should be updated

heshan0131 commented on src/actions/vis-state-actions.js:
```diff
@@ -508,6 +520,76 @@ export function setFeatures(features) {
+/**
+ * Delete the given feature
+ * @memberof visStateActions
+ * @param {string} featureId if of the feature
```
same here

heshan0131 commented on src/components/map/map-control.js:
```diff
@@ -305,18 +319,21 @@ const MapControlFactory = () => {
       mapLayers: PropTypes.object,
-      editor: PropTypes.object
+      editor: PropTypes.object,
+      indexMap: PropTypes.number
```
`indexMap` is defined, but `index` is passed in

macrigiuseppe replied:
actually I noticed we were already passing mapIndex hence we don't need either index now indexMap

heshan0131 replied:
delete `indexMap: PropTypes.number`

heshan0131 commented on src/components/map/map-control.js:
```diff
@@ -305,18 +319,21 @@ const MapControlFactory = () => {
       isSplit: false,
-      top: 0
+      top: 0,
+      indexMap: 0
```
`index` instead of `indexMap`

macrigiuseppe replied:
replace into mapIndex=0

heshan0131 replied:
did you push? it'still showing `indexMap: 0`

macrigiuseppe commented:
@heshan0131 updated

heshan0131 commented on src/actions/vis-state-actions.js:
```diff
@@ -378,6 +378,18 @@ export function enlargeFilter(idx) {
 }
 
+/**
+ * SHow/hide filter feature on map
```
* @memberof visStateActions

heshan0131 commented on src/actions/vis-state-actions.js:
```diff
@@ -508,6 +520,75 @@ export function setFeatures(features) {
+ * It will apply the provide feature as filter to the given layer.
+ * If the given feature is already applied as filter to the layer it  will remove it from  the filter list
+ * @memberof visStateActions
+ * @param {Object} feature
```
layer first, feature second

heshan0131 commented on src/actions/vis-state-actions.js:
```diff
@@ -508,6 +520,75 @@ export function setFeatures(features) {
 
+/**
+ * It will apply the provide feature as filter to the given layer.
+ * If the given feature is already applied as filter to the layer it  will remove it from  the filter list
```
If the given feature is already applied as filter to the layer, it will __remove the layer__ from the filter

heshan0131 commented on src/actions/vis-state-actions.js:
```diff
@@ -508,6 +520,75 @@ export function setFeatures(features) {
+ * Set the current feature to be edited/deleted
+ * @memberof visStateActions
+ * @param {Object} feature
+ * @return {{selectedFeatureId: *, type: ActionTypes.SET_SELECTED_FEATURE}}
```
`@return {{feature: *, type: ActionTypes.SET_SELECTED_FEATURE}}`

heshan0131 commented on src/actions/vis-state-actions.js:
```diff
@@ -508,6 +520,75 @@ export function setFeatures(features) {
+ * Delete the given feature
+ * @memberof visStateActions
+ * @param {Object} feature
+ * @return {{type: ActionTypes.DELETE_FEATURE, featureId: *}}
```
`@return {{type: ActionTypes.DELETE_FEATURE, feature: *}}`

heshan0131 commented on src/actions/vis-state-actions.js:
```diff
@@ -508,6 +520,75 @@ export function setFeatures(features) {
+}
+
+/** Set the map mode
+ * @memberof main
```
`@memberof visStateActions`
`@return {{type: ActionTypes. SET_EDITOR_MODE, mode: *}}`

heshan0131 commented on src/actions/vis-state-actions.js:
```diff
@@ -508,6 +520,75 @@ export function setFeatures(features) {
+}
+
+/**
+ * Toggle editor layer visibility
```
`@memberof visStateActions`

macrigiuseppe commented on examples/demo-app/src/components/map-control/map-control.js:
```diff
@@ -182,7 +182,7 @@ function SampleMapPanel(props) {
 const CustomMapControl = props => (
   <StyledMapControlOverlay>
-    {props.currentSample ? <SampleMapPanel {...props} /> : null}
+    {!props.isExport && props.currentSample ? <SampleMapPanel {...props} /> : null}
```
we don't show info panel when export map

heshan0131 commented on src/components/map-container.js:
```diff
@@ -439,23 +462,25 @@ export default function MapContainerFactory(MapPopover, MapControl) {
       return (
         <StyledMapContainer style={MAP_STYLE.container}>
           <MapControl
+            index={index}
```
we already pass in`mapIndex={this.props.index}`, no need to pass in`index={index}`

macrigiuseppe replied:
yeah, I already removed that one

heshan0131 commented on src/components/map/map-control.js:
```diff
@@ -351,7 +367,8 @@ const MapControlFactory = () => {
         scale,
-        readOnly
+        readOnly,
+        index
```
delete index

macrigiuseppe replied:
yeah done

heshan0131 commented on src/components/map/map-control.js:
```diff
@@ -414,10 +431,11 @@ const MapControlFactory = () => {
             <ActionPanel key={4}>
               <MapDrawPanel
-                isActive={mapDraw.active}
+                isActive={mapDraw.active && mapDraw.activeMapIndex === index}
```
`mapDraw.active && mapDraw.activeMapIndex === mapIndex`

macrigiuseppe replied:
same here

cperbost commented:
> @heshan0131 Added new filter panel with data info:
> 
> <img alt="Screen Shot 2019-08-29 at 4 48 45 PM" width="1792" src="https://user-images.githubusercontent.com/2554552/63950847-08456880-ca7d-11e9-885c-c80fea525a98.png">

Hi @macrigiuseppe , I can't find how to display the "count" (of points inside the polygon) on the layer section. Is does not seem to be implemented in version 2.2.0. Did you remove it?

sulemanbadr commented:
@cperbost did you able to fidn this "count" feature?
